app/models/miq_product_feature.rb

```diff
@@ -153,6 +153,58 @@ def self.seed_feature(hash, features)
     return feature, status
   end
 
+  def self.find_all_by_identifier(features)
+    where(:identifier => features)
+  end
+
+  def self.obj_features
+    @obj_cache ||= begin
+      # create hash with parent identifier and details
+      features = select('*').select(arel_attribute(:parent_identifier).as("parent_identifier")).each_with_object({}) do |f, h|
+        parent_ident = f.parent_identifier
+        h[f.identifier] = {:parent => parent_ident, :children => [], :feature => f}
+      end
+      # populate the children based upon parent identifier
+      features.each do |_, n|
+        if (parent = n[:parent])
+          features[parent][:children] << n[:feature]
+        end
+      end
+      features
+    end
+  end
+
+  def self.obj_feature_all_children(identifier)
+    obj_features[identifier.to_s][:children] if obj_features[identifier.to_s]
+  end
+
+  def self.obj_feature_children(identifier)
+    obj_features[identifier.to_s] && obj_features[identifier.to_s][:children].select { |c| !c.hidden? }
+  end
+
+  def self.obj_feature_parent(identifier)
+    obj_features[identifier.to_s] && obj_features[identifier.to_s][:parent]
+  end
+
+  def self.obj_feature_ancestors(identifier)
+    feature = obj_features[identifier.to_s]
+    ancestors = []
+
+    if feature
```

Comment:
heh. yea. I like iterators over recursion. especially while trying to debug. but up to you